src/theory/strings/theory_strings_rewriter.cpp

```diff
@@ -1754,3 +1729,192 @@ Node TheoryStringsRewriter::collectConstantStringAt( std::vector< Node >& vec, u
+      bool removeComponent = false;
+      Trace("strings-rewrite-debug2") << "stripConstantEndpoints : Compare " << n1[index0] << " " << n2[index1] << ", dir = " << dir << std::endl;
+      if( n1[index0].isConst() ){
+        CVC4::String s = n1[index0].getConst<String>();
+        // overlap is an overapproximation of the number of characters n2[index1] can match in s
+        unsigned overlap = s.size();
+        if( n2[index1].isConst() ){
+          CVC4::String t = n2[index1].getConst<String>();
+          std::size_t ret = s.find( t );
+          if( ret==std::string::npos ) {
+            if( n1.size()==1 ){
+              // can remove everything
+              //   e.g. str.contains( "abc", str.++( "ba", x ) ) --> str.contains( "", str.++( "ba", x ) )
+              removeComponent = true;
+            }else{
+              //check how much overlap there is
+              // This is used to partially strip off the endpoint
+              // e.g. str.contains( str.++( "abc", x ), str.++( "cd", y ) ) --> str.contains( str.++( "c", x ), str.++( "cd", y ) )
+              overlap = r==0 ? s.overlap(t) : t.overlap(s);
+            }
+          }else{
+            Assert( ret<s.size() );
+            // can strip off up to the find position
+            // e.g. str.contains( str.++( "abc", x ), str.++( "b", y ) ) --> str.contains( str.++( "bc", x ), str.++( "b", y ) )
+            overlap = s.size()-ret;
+          }
+        }else if( n2[index1].getKind()==kind::STRING_ITOS ){
+          const std::vector<unsigned>& svec = s.getVec();
+          // can remove up to the first occurrence of a non-digit
+          for( unsigned i=0; i<svec.size(); i++ ){
+            unsigned sindex = r==0 ? i : svec.size()-i;
+            if(String::isDigit(svec[sindex])) {
+              break;
+            }else{
+              // e.g. str.contains( str.++( "a", x ), str.to.int(y) ) --> str.contains( x, str.to.int(y) )
+              overlap--;
+            }
+          }
+        }else{
+          //inconclusive
+        }
+        // process the overlap
+        if( overlap<s.size() ){
+          changed = true;
+          if( overlap==0 ){
+            removeComponent = true;
+          }else{
+            // can drop the prefix (resp. suffix) from the first (resp. last) component
+            if( r==0 ){
+              nb.push_back( NodeManager::currentNM()->mkConst(s.prefix(overlap)) );
+              n1[index0] = NodeManager::currentNM()->mkConst(s.suffix(overlap));
+            }else{
+              ne.push_back( NodeManager::currentNM()->mkConst(s.suffix(overlap)) );
+              n1[index0] = NodeManager::currentNM()->mkConst(s.prefix(overlap));
+            }
+          }
+        }
+      }else if( n1[index0].getKind()==kind::STRING_ITOS ){
+        if( n2[index1].isConst() ){
+          CVC4::String t = n2[index1].getConst<String>();
+          const std::vector<unsigned>& tvec = t.getVec();
+          removeComponent = n1.size()>1;
```

Comment:
`removeComponent = true` should be fine here.

Reply:
This needs to be n1.size()>1. 
When n1.size()==1, then by default we cannot remove the component, but removeComponent is updated true below if there is at least one non-digit.

Reply:
You are right, my bad. Although this makes me think: What if `tvec` is empty and `n1.size() > 1`? Then we remove the `str.to.int` even though the next component in `n2` could be a number. This might not happen in practice but still might be better to always set `removeComponent = false` and then set it to true when we see our first non-digit (this is obviously not needed/easier if we do the more aggressive rewriting that I suggested below).

Reply:
Yes, empty tvec would be a problem, but it will never happen (empty string is removed from concat's before we get to here).  I actually added an assertion !tvec.empty() in my last commit.

Reply:
Anyways, I've refactored this part quite a bit, take a look at the new version and let me know what you think.